packages/pagination/src/add-pagination-body.interceptor.ts

```diff
@@ -0,0 +1,136 @@
+  private readonly pageName: string;
+  private readonly perPageName: string;
+  private readonly defaultLimit: number;
+
+  constructor(options: PaginationBodyInterceptorOptions) {
+    const { defaultLimit = DEFAULT_LIMIT, pageName = 'page', perPageName = 'per_page' } = options;
+
+    this.pageName = pageName;
+    this.perPageName = perPageName;
+    this.defaultLimit = defaultLimit;
+  }
+
+  /**
+   * Interceptor core method
+   * @param context Current request pipeline details
+   * @param next Response stream
+   */
+  public intercept(context: ExecutionContext, next: CallHandler): Observable<PaginatedData<T>> {
+    const req: Request = context.switchToHttp().getRequest();
+    const path: string = req.path;
+    const page: number = !isNaN(Number(req.query[this.pageName])) ? Number(req.query[this.pageName]) : FIRST_PAGE;
+    const limit: number =
+      !isNaN(Number(req.query.limit)) && Number(req.query[this.perPageName]) !== 0
+        ? Number(req.query[this.perPageName])
+        : this.defaultLimit;
+    const filter: unknown = req.query.filter;
+    const sort: unknown = req.query.sort;
+    const project: unknown = req.query.project;
+
+    return next.handle().pipe(
+      // tslint:disable-next-line
+      map((data: DataToPaginate<T>) => {
+        const totalPages: number = Math.ceil(data.totalResources / limit);
+        const nextUri: string | null =
+          page < totalPages ? this.buildUrl(path, page + 1, limit, filter, sort, project) : null;
+        const previousUri: string | null =
+          page > 1 && page <= totalPages ? this.buildUrl(path, page - 1, limit, filter, sort, project) : null;
+        const firstUri: string | null =
+          data.totalResources > 0 ? this.buildUrl(path, FIRST_PAGE, limit, filter, sort, project) : null;
+        const lastUri: string | null =
+          data.totalResources > 0 ? this.buildUrl(path, totalPages, limit, filter, sort, project) : null;
+
+        return {
+          resources: data.resources,
+          pagination: {
+            next: nextUri,
+            previous: previousUri,
+            first: firstUri,
+            last: lastUri,
+            totalPages,
+            totalResources: data.totalResources,
+          },
+        };
+      }),
+    );
+  }
+
+  /**
+   * Build url
+   */
+  public buildUrl = (
+    path: string,
+    page: number,
+    limit: number,
+    filter?: unknown,
+    sort?: unknown,
+    project?: unknown,
+  ): string => {
+    // tslint:disable-next-line
+    let url = `${path}?page=${page}&limit=${limit}`;
+
+    if (filter !== undefined) {
+      url += `&filter=${filter}`;
+    }
+
+    if (sort !== undefined) {
+      url += `&sort=${sort}`;
+    }
+
+    if (project !== undefined) {
+      url += `&project=${project}`;
+    }
+
+    return url;
+  };
+}
+
+/**
+ * Data
+ */
+export interface DataToPaginate<T> {
+  resources: T[];
+  totalResources: number;
+}
```

Comment:
Looks very similar to 
```ts
export interface Pageable<T> {
  resource: T[];
  totalDocs: number;
}
```

Reply:
Should I create a unique interface and export it ? I can use the DataToPaginate name for both of the interceptors

Reply:
Yes you can 🙂 You can also create a new file including common interfaces